### DISCUSSION TITLE
Scenerio for Ops Fetch Violation Error

### DISCUSSION THREAD

**Discussion by d2e** [microsoft/FluidFramework#20082] [Azure Features & Functionality]
I have few questions related to Ops Fetch Violation

- What scenarios typically trigger the Ops Fetch Violation Error?
- What is the potential impact of this error on fluid client users?

**Comment by jatgarg** [comment]
OpsFetchViolation - This error event is recorded when the ops returned by the service are either not contiguous (there is a gap between ops) or the sequence number of first op is not the same as the first sequence number which was requested by the container. 
When the container loads, as part of optimization container just request ops from the snapshot sequence number + 1(seq number at which summary was taken). During that time whatever ops we get from service, cache we apply them to the container on top of snapshot. If there is OpsFetchViolation due to above reasons, only legitimate ops will be applied. But that ops fetching was just an optimization. Later when ops arrive on socket and container realizes that some ops are missing in between the last processed op seq number and the seq number of OP on socket, then it asks for specific missing ops, and it should get it from the service. 
This is not expected to occur again and again. If it causes the container to be stuck and not catch up, then only it is something to be bothered about. In that case, Fluid engineer/AFR team needs to be involved to check the root cause.

**Comment by ashishswami22** [comment]
Is there a pattern in the database which can help us predict this error? e.g. checking if all sequence numbers are present between "content.scribe.lastSummarySequenceNumber" and some end limit (?).

Which field can be used for the end limit of sequence numbers?

**Comment by ashishswami22** [comment]
if there’s more than one condition that causes this error or a similar kind of error?
